display_list/geometry/dl_region.h

```diff
@@ -0,0 +1,65 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_DISPLAY_LIST_GEOMETRY_REGION_H_
+#define FLUTTER_DISPLAY_LIST_GEOMETRY_REGION_H_
+
+#include "third_party/skia/include/core/SkRect.h"
+
+#include <vector>
+
+namespace flutter {
+
+/// Represents a region as a collection of non-overlapping rectangles.
+/// Implements a subset of SkRegion functionality optimized for quickly
+/// converting set of overlapping rectangles to non-overlapping rectangles.
+class DlRegion {
+ public:
+  /// Creates region by bulk adding the rectangles./// Matches
+  /// SkRegion::op(rect, SkRegion::kUnion_Op) behavior.
+  explicit DlRegion(std::vector<SkIRect>&& rects);
+  ~DlRegion();
+
+  /// Returns list of non-overlapping rectangles that cover current region.
+  /// If |deband| is false, each span line will result in separate rectangles,
+  /// closely matching SkRegion::Iterator behavior.
+  /// If |deband| is true, matching rectangles from adjacent span lines will be
+  /// merged into single rectange.
+  std::vector<SkIRect> getRects(bool deband = true) const;
+
+ private:
+  void addRects(std::vector<SkIRect>&& rects);
```

Comment:
Perhaps name this in a way that reminds (internal private) callers that it will mess with the vector. `addRectsUnsafe`? `addMutableRects`?

Reply:
It would be one thing if this were essentially an in/out structure, but it appears to just be a source of data and so someone might assume by default that it doesn't modify that data.

Reply:
`addRects` only accepts a Rvalue and in most cases will not copy without `std::move(rects)`. I'd assume that's a clear indication that it takes ownership of the rects, isn't it?